packages/SwingSet/src/kernel/slogger.js

```diff
@@ -0,0 +1,111 @@
+/* global harden */
+import { assert } from '@agoric/assert';
+
+const IDLE = 'idle';
+const STARTUP = 'startup';
+const DELIVERY = 'delivery';
+
+export function makeDummySlogger(makeConsole) {
+  return harden({
+    addVat: () => 0,
+    vatConsole: () => makeConsole('disabled slogger'),
+    startup: () => () => 0, // returns nop finish() function
+    delivery: () => () => 0,
+    syscall: () => () => 0,
+  });
+}
+
+export function makeSlogger(writeObj) {
+  const write = writeObj ? e => writeObj(harden(e)) : () => 0;
+
+  const vatSlogs = new Map(); // vatID -> vatSlog
+
+  function makeVatSlog(vatID) {
+    let state = IDLE; // or STARTUP or DELIVERY
+    let crankNum;
+    let deliveryNum = 0;
+    let syscallNum;
+
+    function assertOldState(exp, msg) {
+      assert(state === exp, `vat ${vatID} in ${state}, not ${exp}: ${msg}`);
+    }
+
+    function vatConsole(origConsole) {
+      const vc = {};
+      for (const level of ['debug', 'log', 'info', 'warn', 'error']) {
+        vc[level] = (...args) => {
+          origConsole[level](...args);
+          const when = { state, crankNum, vatID, deliveryNum };
+          write({ type: 'console', ...when, level, args });
+        };
+      }
+      return harden(vc);
+    }
+
+    function startup() {
+      // provide a context for console calls during startup
+      assertOldState(IDLE, 'did startup get called twice?');
+      state = STARTUP;
+      function finish() {
+        assertOldState(STARTUP, 'startup-finish called twice?');
+        state = IDLE;
+      }
+      return harden(finish);
+    }
+
+    // kd: kernelDelivery, vd: vatDelivery
+    function delivery(newCrankNum, kd, vd) {
+      assertOldState(IDLE, 'reentrant delivery?');
+      state = DELIVERY;
+      crankNum = newCrankNum;
+      const when = { crankNum, vatID, deliveryNum };
+      write({ type: 'deliver', ...when, kd, vd });
+      deliveryNum += 1;
+      syscallNum = 0;
+
+      // dr: deliveryResult
+      function finish(dr) {
```

Comment:
I'm not wild about this idiom of having each of the log functions return a finish function.  What do we know at the beginning that we don't know at the end?

Reply:
Yeah, I'm on the fence about that approach myself.

The reasons we tell the slog about both delivery-start and delivery-end is so 1: we can record the delivery results somewhere ("the vat died", or "metering reports 14 computrons consumed"), and 2: so the slog knows we're done with a delivery, because any syscalls that happen after that point are illegal (the vat is supposed to be idle).

The delivery results need to be correlated with the delivery start record, and they are matched with a `deliveryNum` (specifically the tuple of `vatID` and `deliveryNum`). But `deliveryNum` is hidden inside the slog code: the kernel doesn't keep track of them. (`crankNum` is recorded too, to let the rendering tool show a total ordering of deliveries, but I'm hesitant to use it for delivery start/finish correlation purposes.. maybe I shouldnt' be). So what `finish` is really closing over is that particular `deliveryNum` value.

But we could implement that by having the `start` return `deliveryNum` instead of `finish()`, and expose a separate `kernelSlog.finishDelivery(vatID, deliveryNum, deliveryResult)` method. The extra arguments seemed redundant to me, an opportunity to get them wrong.

(an earlier iteration exposed distinct `vatSlog` objects for each vat, to remove the `vatID` argument from most of the calls, but finding a place to store it and still access it from the right places seemed messy.. maybe I should revisit that)